cray-mpich +cuda fails to build applications for Spack v0.19.0

Building MPI packages with `cray-mpich-binary+cuda` fails for `v0.19.0` and later of Spack.

The `mpicc` (specifically C language) compiler wrapper fails during the linking stage with the errors like the following:
```
/usr/bin/ld: /user-environment/linux-sles15-zen2/gcc-11.3.0/cray-mpich-binary-8.1.21.1-gcc-4k7oxj3rl75ztwko5s3lamdttacaue4p/lib/libmpi_gtl_cuda.so: undefined reference to `__gxx_personality_v0'
```
`__gxx_personality` is defined in `libstdc++`.

Two manual fixes have been shown to fix the issue:
1. Patching the `mpicc` wrapper to add `-lstdc++`
2. Patching `libmpi_gtl_cuda.so` to resolve its own dependency `patchelf --add-needed libstdc++.so libmpi_gtl_cuda.so`

Of these, the second is more robust, and can be implemented in `cray-mpich-binary/package.py`.

--- a/repo/packages/cray-mpich-binary/package.py
+++ b/repo/packages/cray-mpich-binary/package.py
@@ -52,6 +52,8 @@ class CrayMpichBinary(Package):
     # libfabric.so.1
     depends_on("libfabric@1:", type="link")
 
+    print('----- XXXX installing cray-mpich-binary')
+
     with when("@8.1.21.1-gcc"):
         # libgfortran.so.5
         conflicts("%gcc@:7")
@@ -129,9 +131,14 @@ class CrayMpichBinary(Package):
         for root, _, files in os.walk(self.prefix):
             for name in files:
                 f = os.path.join(root, name)
+                print('----- XXXX testing {}'.format(f))
                 if not self.should_patch(f):
                     continue
+                print('----- XXXX patching {}'.format(f))
                 patchelf("--force-rpath", "--set-rpath", rpath, f, fail_on_error=False)
+                if "libmpi_gtl_cuda.so" in str(f):
+                    print('----- XXXX fixing {}'.format(f))
+                    patchelf("--add-needed", "libstdc++.so", f, fail_on_error=False)
 
     @run_after("install")
     def fixup_compiler_paths(self):
@@ -150,6 +157,8 @@ class CrayMpichBinary(Package):
             gtl_library = "-lmpi_gtl_hsa"
         else:
             gtl_library = ""
+
+        print('----- XXXX patching compilers {}'.format(gtl_library))
 
         filter_file("@@GTL_LIBRARY@@", gtl_library, self.prefix.bin.mpicc, string=True)
         filter_file("@@GTL_LIBRARY@@", gtl_library, self.prefix.bin.mpicxx, string=True)

--- a/repo/packages/cray-mpich-binary/package.py
+++ b/repo/packages/cray-mpich-binary/package.py
@@ -52,8 +52,6 @@ class CrayMpichBinary(Package):
     # libfabric.so.1
     depends_on("libfabric@1:", type="link")
 
-    print('----- XXXX installing cray-mpich-binary')
-
     with when("@8.1.21.1-gcc"):
         # libgfortran.so.5
         conflicts("%gcc@:7")
@@ -131,13 +129,12 @@ class CrayMpichBinary(Package):
         for root, _, files in os.walk(self.prefix):
             for name in files:
                 f = os.path.join(root, name)
-                print('----- XXXX testing {}'.format(f))
                 if not self.should_patch(f):
                     continue
-                print('----- XXXX patching {}'.format(f))
                 patchelf("--force-rpath", "--set-rpath", rpath, f, fail_on_error=False)
+                # The C compiler wrapper can fail because libmpi_gtl_cuda refers to the symbol
+                # __gxx_personality_v0 but wasn't linked against libstdc++.
                 if "libmpi_gtl_cuda.so" in str(f):
-                    print('----- XXXX fixing {}'.format(f))
                     patchelf("--add-needed", "libstdc++.so", f, fail_on_error=False)
 
     @run_after("install")
@@ -157,8 +154,6 @@ class CrayMpichBinary(Package):
             gtl_library = "-lmpi_gtl_hsa"
         else:
             gtl_library = ""
-
-        print('----- XXXX patching compilers {}'.format(gtl_library))
 
         filter_file("@@GTL_LIBRARY@@", gtl_library, self.prefix.bin.mpicc, string=True)
         filter_file("@@GTL_LIBRARY@@", gtl_library, self.prefix.bin.mpicxx, string=True)

